internal/server/handlers.go

```diff
@@ -262,19 +280,31 @@ func (h *handler) ResourceMarshalError(name string, resource proto.Message, err
 	}
 }
 
-func (h *handler) StartNotificationBatch() {
+func (h *handler) StartNotificationBatch(initialResourceVersions map[string]string) {
 	h.lock.Lock()
 	defer h.lock.Unlock()
 
+	// setInitialResourceVersion sets the initial version of resources to filter out unchanged resources.
+	if len(initialResourceVersions) > 0 {
+		h.initialResourceVersions = make(map[string]initialResourceVersion, len(initialResourceVersions))
+		for name, version := range initialResourceVersions {
+			h.initialResourceVersions[name] = initialResourceVersion{version: version}
+		}
+	}
 	h.batchStarted = true
 }
 
 func (h *handler) EndNotificationBatch() {
 	h.lock.Lock()
 	defer h.lock.Unlock()
 
-	h.batchStarted = false
+	for name := range h.initialResourceVersions {
+		if _, ok := h.entries[name]; ok && h.initialResourceVersions[name].received {
+			h.entries[name] = nil
+		}
+	}
```

Comment:
First of all: you're doing an unneeded lookup in this loop. `h.initialResourceVersions[name]` is re-doing the lookup. You should just do `for name, irv := range h.initialResourceVersions` instead, so that you don't rehash `name` every time.

Second, I think this is incorrect. The behavior we want is to send a deletion any resource that is in `h.initialResourceVersions` and *isn't* in `h.entries` (which is what `h.entries[name]` is doing. Here, you are sending deletions for any entry that *does* match, so this is actually a noop since you have this in `Notify`:
```go
	if res, ok := h.initialResourceVersions[name]; ok {
		if r != nil && res.version == r.Version {
			res.received = true
			return
		}
	}
```

Reply:
I think the correct loop here is:
```suggestion
	for name, irv := range h.initialResourceVersions {
		if _, ok := h.entries[name]; !ok && !irv.received {
			h.entries[name] = nil
		}
	}
```